### PR TITLE
feat: add external provisioner daemon helm chart

### PR DESCRIPTION
**Pull request by spikecurtis**
Adds a new helm chart for external provisioner daemons.

Refactors existing helm chart to pull common templates into a `libcoder` library chart.  This is in the first commit, so you might find it easier to review commit-by-commit.

This PR doesn't ship the new helm chart to our repository.  I'll follow up in another PR since this one is big already.

### CODE REVIEW

**Review by mtojek — COMMENTED** [review]
I did some QA to verify the PR.

1. Maybe you have planned to introduce it as a follow-up, but it might be worth flushing out updated docs aligned with these changes. Alternatively, you can write down some CONTRIBUTING for development purposes.
2. Is there an option to add some prechecks to block starting the deployment if env/properties are missing?

For instance:

```
kubectl create secret generic coder-provisioner-psk -n coder --from-literal=psk=aaaaaa
helm install -f values.yaml -n coder --set coder.image.tag=latest coder-provisioner .

kubectl get pods -n coder
NAME                                READY   STATUS             RESTARTS       AGE
coder-db-postgresql-0               1/1     Running            0              22m
coder-provisioner-67c786b98-8jkts   0/1     CrashLoopBackOff   7 (3m1s ago)   14m

kubectl logs coder-provisioner-67c786b98-8jkts -n coder
running command "coder provisionerd start": You are not logged in. Try logging in using 'coder login <url>'.:
    github.com/coder/coder/cli.init
        /home/runner/actions-runner/_work/coder/coder/cli/root.go:74
```

File: helm/coder/charts/libcoder-0.1.0.tgz
[the remark this answers is not in the document]
Comment (spikecurtis) [reply helm/coder/charts/libcoder-0.1.0.tgz:1]: If we don't then we need to do a `helm dependency build` before using the chart.  The prior chart was usable from a clean git checkout, so I thought we'd like to keep that property.

File: helm/provisioner/tests/testdata/missing_values.yaml
[the remark this answers is not in the document]
Comment (spikecurtis) [reply helm/provisioner/tests/testdata/missing_values.yaml:1]: Yeah, it's testing that running with no values set produces an error

File: helm/provisioner/tests/chart_test.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+// These tests run `helm template` with the values file specified in each test
+// and compare the output to the contents of the corresponding golden file.
+// All values and golden files are located in the `testdata` directory.
+// To update golden files, run `go test . -update`.
+
+// UpdateGoldenFiles is a flag that can be set to update golden files.
+var UpdateGoldenFiles = flag.Bool("update", false, "Update golden files")
+
+var TestCases = []TestCase{
```
[the remark this answers is not in the document]
Comment (spikecurtis) [reply helm/provisioner/tests/chart_test.go:26]: Probably don't need to be; I copied from the existing coder chart.

File: helm/coder/README.md
[the remark this answers is not in the document]
Comment (spikecurtis) [reply helm/coder/README.md:1]: yes.  The existing coder chart is refactored to pull out the `libcoder` stuff so I could reuse it for the provisionerd chart, but functionally, it should be the same.  The only change is to add a new optional field to `values.yaml` to set the PSK secret.

**Comment by spikecurtis** [comment]
> Maybe you have planned to introduce it as a follow-up, but it might be worth flushing out updated docs aligned with these changes. Alternatively, you can write down some CONTRIBUTING for development purposes.

Yeah, good point.  I know I need to write docs for the new chart, but didn't mention it in what was missing.  I'll address in a follow on PR.

**Comment by spikecurtis** [comment]
> Is there an option to add some prechecks to block starting the deployment if env/properties are missing?
>For instance:
>
>kubectl create secret generic coder-provisioner-psk -n coder --from-literal=psk=aaaaaa
>helm install -f values.yaml -n coder --set coder.image.tag=latest coder-provisioner .
>
>kubectl get pods -n coder
>NAME                                READY   STATUS             RESTARTS       AGE
>coder-db-postgresql-0               1/1     Running            0              22m
>coder-provisioner-67c786b98-8jkts   0/1     CrashLoopBackOff   7 (3m1s ago)   14m
>
>kubectl logs coder-provisioner-67c786b98-8jkts -n coder
>running command "coder provisionerd start": You are not logged in. Try logging in using 'coder login <url>'.:
>    github.com/coder/coder/cli.init
>        /home/runner/actions-runner/_work/coder/coder/cli/root.go:74

I think what's happening in your example isn't that any fields are missing --- the only required field is the image tag,  which you've set.  The problem is that `latest` is v2.0.0, which is prior to #8877 that adds the PSK.  So `coder provisionerd start` is looking for a session token and sees that you don't have one.  It doesn't understand the PSK environment variable we set.

File: helm/libcoder/Chart.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+apiVersion: v2
```
[the remark this answers is not in the document]
Comment (spikecurtis) [reply helm/libcoder/Chart.yaml:1]: I'll add one to provisioner anyway when I write the docs PR.  The README.md is pretty basic and mainly points to our docs.

**Comment by matifali** [comment]
@spikecurtis: For testing, you can use ghcr.io/coder/coder-preview:main as we build an image for each commit to the `main`.

File: helm/coder/charts/libcoder-0.1.0.tgz
[the remark this answers is not in the document]
Comment (mtojek) [reply helm/coder/charts/libcoder-0.1.0.tgz:1]: Thanks for clarifying. SGTM.

File: helm/libcoder/Chart.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mtojek) [reply helm/libcoder/Chart.yaml:1]: I'm fine with this order as long as it will not invalidate any instructions in Coder docs.

**Comment by mtojek** [comment]
Thanks for the hint, @matifali!

I managed to start the provisioner:

```
kubectl get pods -n coder
NAME                                READY   STATUS    RESTARTS   AGE
coder-589cdbd87c-cw9fs              1/1     Running   0          21m
coder-db-postgresql-0               1/1     Running   0          3h29m
coder-provisioner-7f66b7b5c-dv2mk   1/1     Running   0          9m55s
➜  provisioner git:(spike/8243-helm-ext-provisioner) ✗ kubectl logs coder-provisioner-7f66b7b5c-dv2mk -n coder -f
2023-08-09 12:09:32.815 [info]  starting provisioner daemon  tags={}
```

Few suggestions:
* I started with zero-knowledge about necessary configuration and saw a couple of error logs:

```
2023-08-09 11:57:43.619 [warn]  coderd client failed to dial  error="GET http://coder.coder.svc.cluster.local/api/v2/organizations/00000000-0000-0000-0000-000000000000/provisionerdaemons/serve?provisioner=terraform: unexpected status code 403: You aren't allowed to create provisioner daemons"
2023-08-09 11:57:53.628 [warn]  coderd client failed to dial  error="GET http://coder.coder.svc.cluster.local/api/v2/organizations/00000000-0000-0000-0000-000000000000/provisionerdaemons/serve?provisioner=terraform: unexpected status code 403: You aren't allowed to create provisioner daemons"
2023-08-09 11:58:03.634 [warn]  coderd client failed to dial  error="GET http://coder.coder.svc.cluster.local/api/v2/organizations/00000000-0000-0000-0000-000000000000/provisionerdaemons/serve?provisioner=terraform: unexpected status code 403: You aren't allowed to create provisioner daemons"
2023-08-09 11:58:13.643 [warn]  coderd client failed to dial  error="GET http://coder.coder.svc.cluster.local/api/v2/organizations/00000000-0000-0000-0000-000000000000/provisionerdaemons/serve?provisioner=terraform: unexpected status code 403: You aren't allowed to create provisioner daemons"
2023-08-09 11:58:23.666 [warn]  coderd client failed to dial  error="GET http://coder.coder.svc.cluster.local/api/v2/organizations/00000000-0000-0000-0000-000000000000/provisionerdaemons/serve?provisioner=terraform: unexpected status code 403: External provisioner daemons is an Enterprise feature. Contact sales!"
2023-08-09 11:58:33.678 [warn]  coderd client failed to dial  error="GET http://coder.coder.svc.cluster.local/api/v2/organizations/00000000-0000-0000-0000-000000000000/provisionerdaemons/serve?provisioner=terraform: unexpected status code 403: External provisioner daemons is an Enterprise feature. Contact sales!"
2023-08-09 11:58:43.689 [warn]  coderd client failed to dial  error="GET http://coder.coder.svc.cluster.local/api/v2/organizations/00000000-0000-0000-0000-000000000000/provisionerdaemons/serve?provisioner=terraform: unexpected status code 403: External provisioner daemons is an Enterprise feature. Contact sales!"
2023-08-09 11:58:53.697 [warn]  coderd client failed to dial  error="GET http://coder.coder.svc.cluster.local/api/v2/organizations/00000000-0000-0000-0000-000000000000/provisionerdaemons/serve?provisioner=terraform: unexpected status code 403: External provisioner daemons is an Enterprise feature. Contact sales!"
2023-08-09 11:59:03.709 [warn]  coderd client failed to dial  error="GET http://coder.coder.svc.cluster.local/api/v2/organizations/00000000-0000-0000-0000-000000000000/provisionerdaemons/serve?provisioner=terraform: unexpected status code 403: External provisioner daemons is an Enterprise feature. Contact sales!"
2023-08-09 11:59:13.713 [warn]  coderd client failed to dial  error="GET http://coder.coder.svc.cluster.local/api/v2/organizations/00000000-0000-0000-0000-000000000000/provisionerdaemons/serve?provisioner=terraform: unexpected status code 403: External provisioner daemons is an Enterprise feature. Contact sales!"
2023-08-09 11:59:23.719 [warn]  coderd client failed to dial  error="GET http://coder.coder.svc.cluster.local/api/v2/organizations/00000000-0000-0000-0000-000000000000/provisionerdaemons/serve?provisioner=terraform: unexpected status code 403: External provisioner daemons is an Enterprise feature. Contact sales!"
2023-08-09 11:59:33.731 [warn]  coderd client failed to dial  error="GET http://coder.coder.svc.cluster.local/api/v2/organizations/00000000-0000-0000-0000-000000000000/provisionerdaemons/serve?provisioner=terraform: unexpected status code 403: External provisioner daemons is an Enterprise feature. Contact sales!"
```

After I set the license, the log stream stopped. It might be useful for ops to leave an extra info message confirming that provisioner works correctly, and it is up and running (healthcheck in logs?).

* Should I be able to connect the provisoner with coderd, so that new provisioner can pick up tasks? So far, I can see that "coder" deployment uses its embedded provisioner, and I'm wondering if this is in scope of this PR. I modified `pskSecretName` to match `coder-provisioner-psk`, not sure if this is required?

**Comment by spikecurtis** [comment]
> Should I be able to connect the provisoner with coderd, so that new provisioner can pick up tasks? So far, I can see that "coder" deployment uses its embedded provisioner, and I'm wondering if this is in scope of this PR. I modified `pskSecretName` to match `coder-provisioner-psk`, not sure if this is required?

Yes, the new provisioner should be able to pick up tasks, but it will race with the provisionerds that are in-process with Coderd to acquire jobs.  You can create a lot of jobs in parallel, or you could start Coderd without in-process provisionerd by setting

```
    - name: CODER_PROVISIONER_DAEMONS
      value: "0"
```
in the `coder.env` value.

**Comment by spikecurtis** [comment]
> After I set the license, the log stream stopped. It might be useful for ops to leave an extra info message confirming that provisioner works correctly, and it is up and running (healthcheck in logs?).

Yeah, I agree it's useful to drop an INFO log when we successfully connect.  I'll add one.

**Review by mtojek — APPROVED** [review]
Helm chart works correctly. Nice job 👍 

This could be potentially an issue:

```
[
   {
      "id":"6221b964-d2ed-4faa-9864-ad077d35a4de",
      "created_at":"2023-08-10T07:39:08.633353Z",
      "updated_at":{
         "Time":"0001-01-01T00:00:00Z",
         "Valid":false
      },
      "name":"priceless_shaw4",
      "provisioners":[
         "terraform"
      ],
      "tags":{
         
      }
   },
   {
      "id":"48bdb066-bc75-4365-b655-4d7e3bc71c7e",
      "created_at":"2023-08-10T07:49:20.504444Z",
      "updated_at":{
         "Time":"0001-01-01T00:00:00Z",
         "Valid":false
      },
      "name":"reverent_johnson2",
      "provisioners":[
         "terraform"
      ],
      "tags":{
         
      }
   },
   {
      "id":"3465c1ac-c6d5-4ba4-8442-834d902a8a1e",
      "created_at":"2023-08-10T07:49:30.330417Z",
      "updated_at":{
         "Time":"0001-01-01T00:00:00Z",
         "Valid":false
      },
      "name":"jolly_feistel6",
      "provisioners":[
         "terraform"
      ],
      "tags":{
         
      }
   },
   {
      "id":"820684dd-35bf-4bc7-bfd4-c6c565eefb15",
      "created_at":"2023-08-10T07:49:35.860361Z",
      "updated_at":{
         "Time":"0001-01-01T00:00:00Z",
         "Valid":false
      },
      "name":"beautiful_sutherland5",
      "provisioners":[
         "terraform"
      ],
      "tags":{
         
      }
   },
   {
      "id":"dc127705-726d-4bfb-8923-3ce40fae05fb",
      "created_at":"2023-08-10T07:51:45.435515Z",
      "updated_at":{
         "Time":"0001-01-01T00:00:00Z",
         "Valid":false
      },
      "name":"upbeat_heisenberg8",
      "provisioners":[
         "terraform"
      ],
      "tags":{
         
      }
   },
   {
      "id":"cdbb9d36-4e63-4e21-a6a6-9d97d582244d",
      "created_at":"2023-08-10T07:51:45.687759Z",
      "updated_at":{
         "Time":"0001-01-01T00:00:00Z",
         "Valid":false
      },
      "name":"zealous_antonelli3",
      "provisioners":[
         "terraform"
      ],
      "tags":{
         
      }
   },
   {
      "id":"820e9694-e512-45f5-98c5-b78a729b89f4",
      "created_at":"2023-08-10T07:51:45.74739Z",
      "updated_at":{
         "Time":"0001-01-01T00:00:00Z",
         "Valid":false
      },
      "name":"amazing_williamson9",
      "provisioners":[
         "terraform"
      ],
      "tags":{
         
      }
   },
   {
      "id":"7982f865-f8e3-4337-969b-789926fd3d13",
      "created_at":"2023-08-10T07:51:46.056332Z",
      "updated_at":{
         "Time":"0001-01-01T00:00:00Z",
         "Valid":false
      },
      "name":"distracted_robinson5",
      "provisioners":[
         "terraform"
      ],
      "tags":{
         
      }
   },
   {
      "id":"12ec2e26-ee1c-4b8a-a104-09b60ba58059",
      "created_at":"2023-08-10T07:51:46.930575Z",
      "updated_at":{
         "Time":"0001-01-01T00:00:00Z",
         "Valid":false
      },
      "name":"focused_kirch8",
      "provisioners":[
         "terraform"
      ],
      "tags":{
         
      }
   },
   {
      "id":"0f9b8fad-e013-4cde-8653-174f15c57bda",
      "created_at":"2023-08-10T07:53:34.58902Z",
      "updated_at":{
         "Time":"0001-01-01T00:00:00Z",
         "Valid":false
      },
      "name":"eager_goldstine7",
      "provisioners":[
         "terraform"
      ],
      "tags":{
         
      }
   },
   {
      "id":"5e904073-a3b3-4ed2-89e2-33cbf0cca1f4",
      "created_at":"2023-08-10T07:53:34.754871Z",
      "updated_at":{
         "Time":"0001-01-01T00:00:00Z",
         "Valid":false
      },
      "name":"cranky_bardeen3",
      "provisioners":[
         "terraform"
      ],
      "tags":{
         
      }
   },
   {
      "id":"b1053a9d-6b50-4631-8353-bae5d668fd01",
      "created_at":"2023-08-10T07:53:34.846211Z",
      "updated_at":{
         "Time":"0001-01-01T00:00:00Z",
         "Valid":false
      },
      "name":"vibrant_ptolemy7",
      "provisioners":[
         "terraform"
      ],
      "tags":{
         
      }
   },
   {
      "id":"d540e06a-8813-46db-90e8-af28afff1a49",
      "created_at":"2023-08-10T07:53:35.265534Z",
      "updated_at":{
         "Time":"0001-01-01T00:00:00Z",
         "Valid":false
      },
      "name":"vigorous_mclean2",
      "provisioners":[
         "terraform"
      ],
      "tags":{
         
      }
   },
   {
      "id":"d5f83605-203a-4fa3-ad11-a1e1154674a3",
      "created_at":"2023-08-10T07:53:35.694644Z",
      "updated_at":{
         "Time":"0001-01-01T00:00:00Z",
         "Valid":false
      },
      "name":"friendly_shirley6",
      "provisioners":[
         "terraform"
      ],
      "tags":{
         
      }
   }
]
```

I uninstalled & installed provisioner instances a couple of times, and they are all exposed over API. Considering the daily pattern of scale up & down, we may blow up the table really quickly. I'm wondering if we need to clean up old daemon entries.

**Comment by spikecurtis** [comment]
> I uninstalled & installed provisioner instances a couple of times, and they are all exposed over API. Considering the daily pattern of scale up & down, we may blow up the table really quickly. I'm wondering if we need to clean up old daemon entries.

Yeah, the provisioner daemons table in the DB is a roach motel.  We only ever add to it.  It's also not clear to me what the purpose of the table is.  Tracking in #9015

**Pull request merged**